File: read/src/main/scala/io/github/jarol/azure/search/spark/connector/read/SearchDocumentEncoder.scala
package io.github.jarol.azure.search.spark.connector.read

import com.azure.search.documents.SearchDocument
import com.azure.search.documents.indexes.models.SearchField
import io.github.jarol.azure.search.spark.connector.core.JavaScalaConverters
import io.github.jarol.azure.search.spark.connector.core.schema.conversion.input.SearchEncoder
import io.github.jarol.azure.search.spark.connector.core.schema.conversion.{SchemaViolationException, SearchIndexColumn}
import org.apache.spark.sql.catalyst.InternalRow
import org.apache.spark.sql.types.StructType

/**
 * Encoder for translating a [[SearchDocument]] into an [[InternalRow]]
 * @param indexColumnToEncoders encoders to apply in order to extract row values
 */

case class SearchDocumentEncoder(private val indexColumnToEncoders: Map[SearchIndexColumn, SearchEncoder])
  extends (SearchDocument => InternalRow) {

  private lazy val sortedEncoders: Seq[(String, SearchEncoder)] = indexColumnToEncoders.toSeq.sortBy {
    case (k, _) => k.index()
  }.map {
    case (k, v) => (k.name(), v)
  }

  override def apply(v1: SearchDocument): InternalRow = {

    val values: Seq[Any] = sortedEncoders.map {
      case (name, encoder) => encoder.apply(v1.get(name))
    }

    InternalRow.fromSeq(values)
  }
}

object SearchDocumentEncoder {

  /**
   * Safely create a document encoder instance
   * @param schema target Dataframe schema
   * @param searchFields input Search fields
   * @return either the encoder or a [[SchemaViolationException]]
   */

  final def safeApply(
                       schema: StructType,
                       searchFields: Seq[SearchField]
                     ): Either[SchemaViolationException, SearchDocumentEncoder] = {

    EncodersSupplier.get(
      schema,
      searchFields
    ).left.map {
      // Map left side
      v => new SchemaViolationException(
        JavaScalaConverters.seqToList(v)
      )
    }.right.map {
      // Create encoder
      v => SearchDocumentEncoder(v)
    }
  }
}

File: connector/src/main/java/io/github/dejarol/azure/search/spark/connector/read/encoding/SearchDocumentEncoder.java
package io.github.dejarol.azure.search.spark.connector.read.encoding;

import com.azure.search.documents.SearchDocument;
import org.apache.spark.sql.catalyst.InternalRow;

import java.io.Serializable;

/**
 * Interface for defining encoders that should convert {@link com.azure.search.documents.SearchDocument}(s)
 * to {@link org.apache.spark.sql.catalyst.InternalRow}(s). They must be serializable as they must be propagated
 * to each read partition
 */

@FunctionalInterface
public interface SearchDocumentEncoder
        extends Serializable {

    /**
     * Converts the given document to a Spark internal row
     * @param document a document retrieved from a Search index
     * @return a Spark internal row that contains document values
     */

    InternalRow apply(
            SearchDocument document
    );
}
